specification/hugr.md

```diff
@@ -1168,44 +1168,78 @@ The new hugr is then derived as follows:
-    μ<sub>out</sub>;
+A `NewOutEdgeSpec` specifies an edge inserted from a new node to an existing node.
+It contains the following fields:
 
-3.  for each node n in top(G), adding a hierarchy edge from t(n) to n,
-    placing n in the first position among children of t(n) if n is in
-    Init and in the second position if n is in Term;
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `TgtNode`: the target node of the new edge.
+  - `TgtPort`: (for `Value` and `Static` edges only) the target port (whose
+    existing incoming edge will be removed).
+  - `FromPos` (for `Value`, `Static` and `ControlFlow` edges only) the desired
+    position among the outgoing ports from the new node.
 
-4.  for each node n in bot(G), and for each child m of b(n), adding a
-    hierarchy edge from n to m (replacing m’s existing parent edge)
+The `Replace` method takes as input:
 
-5.  removing all nodes in R
+  - a set $S$ of mutually-separated nodes in $\Gamma$;
+  - a partial hugr $G$;
+  - a map $T : \mathrm{top}(G) \to \Gamma \setminus S^*$ whose image consists of container nodes;
+  - a map $B : \mathrm{bot}(G) \to S^\*$ whose image consists of container nodes, such that $B(x)$
+    is separated from $B(y)$ unless $x = y$. Let $X$ be the set of children
+    of nodes $B(y)$ for $y \in \mathrm{bot}(G)$, and $R = S^\* \setminus X^\*$.
+  - a list $\mu\_\textrm{inp}$ of pairs $(n, \sigma\_\mathrm{inp})$ where $n$ is a node in $G$ and $\sigma\_\mathrm{inp}$ is a `NewInpEdgeSpec`;
+  - a list $\mu\_\textrm{out}$ of pairs $(n, \sigma\_\mathrm{out})$ where $n$ is a node in $G$ and $\sigma\_\mathrm{out}$ is a `NewOutEdgeSpec`.
 
-6.  If any edges inserted in step 2 are inter-graph (i.e DFG.
-    non-sibling), inserting any `Order` edges required to validate them.
+The new hugr is then derived as follows:
+
+1.  Make a copy in $\Gamma$ of all the nodes in $G$, and all edges between them.
+2.  For each $(n, \sigma\_\mathrm{inp}) \in \mu\_\textrm{inp}$, insert a new edge going into the new
+    copy of $n$ according to the specification $\sigma\_\mathrm{inp}$. (Note: some
+    bookkeeping will be required to ensure the eventual input port orderings of newly added nodes
+    are correct.)
+3.  For each $(n, \sigma\_\mathrm{out}) \in \mu\_\textrm{out}$, insert a new edge going out of the new
+    copy of $n$ according to the specification $\sigma\_\mathrm{out}$. (Note: it should
+    be verified that the predecessor node of the existing target port is in $R$,
+    and so will be removed at step 6.)
+4.  For each $(n, t = T(n))$, append the copy of $n$ to the list
+    of children of $t$ (adding a hierachy edge from $t$ to $n$).
+5.  For each node $(n, b = B(n))$ and for each child $m$ of $b$, replace $m$ with
+    the new copy of $n$ in the list of children of $b$ (replacing the existing hierarchy edge).
+6.  Remove all nodes in $R$ and edges adjoining them.
+7.  If any edges inserted in step 2 are inter-graph (i.e. DFG,
```

Comment:
"inter-graph" -> "value/static edges with locality `ext`"
But, I wonder if we could instead require these to be included in the `\mu_\textrm{in}` / `\mu_\textrm{out}`?

Reply:
Yes I think we can, that's a good idea.